tensorflow/contrib/data/kernels/unordered_merge_dataset_op.cc

```diff
@@ -0,0 +1,218 @@
+   public:
+    explicit Dataset(OpKernelContext* ctx,
+                     const std::vector<DatasetBase*>& inputs)
+        : GraphDatasetBase(ctx), inputs_(inputs) {
+      const auto& input = inputs_[0];
+      for (DataType dt : input->output_dtypes()) {
+        output_dtypes_.push_back(dt);
+      }
+      output_shapes_.insert(output_shapes_.end(),
+                            input->output_shapes().begin(),
+                            input->output_shapes().end());
+      for (const auto& input : inputs_) {
+        input->Ref();
+      }
+    }
+
+    ~Dataset() override {
+      for (const auto& input : inputs_) {
+        input->Unref();
+      }
+    }
+
+    std::unique_ptr<IteratorBase> MakeIterator(
+        const string& prefix) const override {
+      return std::unique_ptr<IteratorBase>(
+          new Iterator({this, strings::StrCat(prefix, "::UnorderedMerge")}));
+    }
+
+    const DataTypeVector& output_dtypes() const override {
+      return output_dtypes_;
+    }
+
+    const std::vector<PartialTensorShape>& output_shapes() const override {
+      return output_shapes_;
+    }
+
+    string DebugString() override { return "UnorderedMergeDatasetOp::Dataset"; }
+
+   protected:
+    Status AsGraphDefInternal(OpKernelContext* ctx, DatasetGraphDefBuilder* b,
+                              Node** output) const override {
+      std::vector<Node*> input_graph_nodes;
+      input_graph_nodes.reserve(inputs_.size());
+      for (const auto& input : inputs_) {
+        Node* input_node;
+        TF_RETURN_IF_ERROR(b->AddParentDataset(ctx, input, &input_node));
+        input_graph_nodes.emplace_back(input_node);
+      }
+      TF_RETURN_IF_ERROR(b->AddDataset(
+          this, {}, {std::make_pair(0, input_graph_nodes)}, {}, output));
+      return Status::OK();
+    }
+
+   private:
+    class Iterator : public DatasetIterator<Dataset> {
+     public:
+      explicit Iterator(const Params& params)
+          : DatasetIterator<Dataset>(params) {
+        input_impls_.reserve(params.dataset->inputs_.size());
+        size_t idx = 0;
+        for (const auto& input : params.dataset->inputs_) {
+          input_impls_.emplace_back(input->MakeIterator(
+              strings::StrCat(params.prefix, "[", idx++, "]")));
+        }
+      }
+
+      Status GetNextInternal(IteratorContext* ctx,
+                             std::vector<Tensor>* out_tensors,
+                             bool* end_of_sequence) override {
+        mutex_lock l(mu_);
+        if (input_impls_.empty()) {
+          *end_of_sequence = true;
+          return Status::OK();
+        }
+        out_tensors->clear();
+        out_tensors->reserve(1);
+        int orig_input_start = input_start;
+        do {
+          const auto& input_impl = input_impls_[input_start];
+          std::vector<Tensor> input_tensors;
+          TF_RETURN_IF_ERROR(
+              input_impl->GetNext(ctx, &input_tensors, end_of_sequence));
```

Comment:
The implementation here is deterministic, which means that it's equivalent to `tf.contrib.data.sample_from_datasets(tf.data.Dataset.range(len(datasets)).repeat(None).map(lambda i: tf.one_hot(i, len(datasets))), datasets)`.

I think the original spirit of the feature request was to implement a non-deterministic merge between multiple datasets, similar to `tf.contrib.data.parallel_interleave(..., sloppy=True)`, such that elements would be produced from any available input as soon as they were ready. Can you try implementing that behavior instead?

Reply:
The API tf.contrib.data.sample_from_datasets looks like a new one and maybe satisfy my original request -- my request **allows** (not requires) non-deterministic behavior -- if it's fast enough. I think I have to try evaluating it.